system/mp/mpcomm.cpp

```diff
@@ -700,33 +701,76 @@ void traceSlowReadTms(const char *msg, ISocket *sock, void *dst, size32_t minSiz
     dbgassertex(timeoutChkIntervalMs < timeoutMs);
     StringBuffer epStr;
     CCycleTimer readTmsTimer;
-    unsigned intervalTimeoutMs = timeoutChkIntervalMs;
+    unsigned intervalTimeoutMs = 500;
+    CCycleTimer intvlTimer;
+
+    // legacy client sends minSize, recent client sends maxSize
+    // if read < maxSize, keep trying for maxSize, but if its exactly minSize
+    // somewhat quickly (without waiting full timeout) settle for minSize ...
+
+    if (intervalTimeoutMs > timeoutChkIntervalMs)
+        intervalTimeoutMs = timeoutChkIntervalMs;
+
+    unsigned minTimeIters = 5000 / intervalTimeoutMs;
+
+    sizeRead = 0;
+
+    unsigned numIters = 0;
+    size32_t maxRead = maxSize;
     for (;;)
     {
         try
         {
-            sock->readtms(dst, minSize, maxSize, sizeRead, intervalTimeoutMs);
-            break;
+            size32_t amtRead = 0;
+            sock->readtms((char *)dst+sizeRead, 0, maxRead, amtRead, intervalTimeoutMs);
+            sizeRead += amtRead;
+            if (sizeRead == maxSize)
+                break;
+            maxRead -= amtRead;
         }
         catch (IJSOCK_Exception *e)
         {
             if (JSOCKERR_graceful_close == e->errorCode())
+            {
+                e->Release();
                 return;
+            }
             else if (JSOCKERR_timeout_expired != e->errorCode())
                 throw;
+            // interval read timed out ...
+            if (sizeRead == minSize)
+            {
+                if (numIters++ >= minTimeIters) // 5 sec max addl wait
```

Comment:
I am happy to merge as is, but I think the following would be clearer:

```
if (firstReadTime == 0)
   firstReadTime = elapsedMs;
else if ((elapsedMs - firstReadTime) >= 5000)
{
   e->Release();
   break;
}
```
remove numIters and move line 749 up before this test.

Reply:
ok, it is more clear this way, changed